mapping_parenting_tech/analysis/prototyping/play_store/get_play_store_apps.py

```diff
@@ -7,141 +7,413 @@
+
+    The function returns a list of dict objects, which it also saves on-the-fly in the target file. `username` and
+    `userimage` are removed from the reviews that are returned and saved for data privacy.
+
+    `filename` is also used as the basis of a logfile `[filename].log`. If the function fails during run-time, it
+    will resume where it left off, provided the same filename is given.
+    """
+
+    target_file = PROJECT_DIR / "outputs/data" / filename
+    file_root = Path(target_file).stem
+    log_file = PROJECT_DIR / "outputs/data" / (file_root + ".log")
 
-logging.info(f"Retrieved {running_total} reviews")
+    # Field names being used in the CSV file
+    field_names = [
+        "content",
+        "score",
+        "thumbsUpCount",
+        "reviewCreatedVersion",
+        "at",
+        "replyContent",
+        "repliedAt",
+        "reviewId",
+        "appId",
+    ]
+
+    # Does the target file (filename) already exist? If so, get a list of the ids from the reviews already in there
+    if target_file.exists():
+        with open(target_file, "rt", newline="") as csv_file:
+            fetched_reviews = csv.DictReader(csv_file, fieldnames=field_names)
+            existing_review_ids = [r["reviewId"] for r in fetched_reviews]
+
+        # Is there a log file - this contains details of reviews downloaded for each app given, provided it's being saved to
+        # the same filename as used originally
+        if log_file.exists():
+            log_file_handle = open(log_file, "rb")
+            log_info = pickle.load(log_file_handle)
+        else:
+            log_file_handle = open(log_file, "wb")
+            log_info = {}
+        log_file_handle.close()
+
+    # If not, create empty new file for reviews and log
+    else:
+        with open(target_file, "wt", newline="") as csv_file:
+            csv_writer = csv.DictWriter(csv_file, fieldnames=field_names)
+            csv_writer.writeheader()
+
+        fetched_reviews = []
+        existing_review_ids = []
+        log_file_handle = open(log_file, "wb")
+        log_info = {}
+        log_file_handle.close()
+
+    # tqdm(app_id_list, desc="Retrieving app reviews"):
+    for app_id in app_id_list:
+        # reset variables ahead of downloading an app's reviews
+        first_pass = True
+        keep_going = True
+        more_to_get = True
+        continuation_token = None
+        review_fetch = []
+        app_review_count = 0
+
+        # Is the app in the logfile? If so, is it completed?
+        if app_id in log_info:
+            if log_info[app_id]["completed"]:
+                keep_going = False
+            else:
+                continuation_token = log_info[app_id]["continuation_token"]
+                first_pass = False
+                more_to_get = True
+                log_dump = {app_id: log_info[app_id]}
+
+        # if we're forcing a download, then at least download the first set of reviews
+        if force_download:
```

Comment:
Apologies, I'm not sure what `force_download` is doing here, because `keep_going` and `first_pass` have already been reset to True in lines 316-317 ?

Reply:
@beingkk `keep_going` can be set to `False` in line 326, so this overwrites that. But, re-reading, re-assigning `first_pass` in unnecessary as it can't get here unless it's `True`